portkey_ai/api_resources/base_client.py

```diff
@@ -376,3 +377,351 @@ def _make_status_error_from_response(
+        headers_dict = self._merge_mappings(
+            self._default_headers, option_headers, self.custom_headers
+        )
+        headers = httpx.Headers(headers_dict)
+        return headers
+
+    def _merge_mappings(
+        self,
+        *args,
+    ) -> Dict[str, Any]:
+        """Merge two mappings of the given type
+        In cases with duplicate keys the second mapping takes precedence.
+        """
+        mapped_headers = {}
+        for i in args:
+            mapped_headers.update(i)
+        return mapped_headers
+
+    def is_closed(self) -> bool:
+        return self._client.is_closed
+
+    async def close(self) -> None:
+        """Close the underlying HTTPX client.
+
+        The client will *not* be usable after this.
+        """
+        await self._client.aclose()
+
+    async def __aenter__(self: Any) -> Any:
+        return self
+
+    async def __aexit__(
+        self,
+        exc_type: Optional[BaseException],
+        exc: Optional[BaseException],
+        exc_tb: Optional[TracebackType],
+    ) -> None:
+        await self.close()
+
+    async def _build_request(self, options: Options) -> httpx.Request:
+        headers = self._build_headers(options)
+        params = options.params
+        json_body = options.json_body
+        request = self._client.build_request(
+            method=options.method,
+            url=options.url,
+            headers=headers,
+            params=params,
+            json=json_body,
+            timeout=options.timeout,
+        )
+        return request
+
+    @overload
+    async def _request(
+        self,
+        *,
+        options: Options,
+        stream: Literal[False],
+        cast_to: Type[ResponseT],
+        stream_cls: Type[AsyncStreamT],
+    ) -> ResponseT:
+        ...
+
+    @overload
+    async def _request(
+        self,
+        *,
+        options: Options,
+        stream: Literal[True],
+        cast_to: Type[ResponseT],
+        stream_cls: Type[AsyncStreamT],
+    ) -> AsyncStreamT:
+        ...
+
+    @overload
+    async def _request(
+        self,
+        *,
+        options: Options,
+        stream: bool,
+        cast_to: Type[ResponseT],
+        stream_cls: Type[AsyncStreamT],
+    ) -> Union[ResponseT, AsyncStreamT]:
+        ...
+
+    async def _request(
+        self,
+        *,
+        options: Options,
+        stream: bool,
+        cast_to: Type[ResponseT],
+        stream_cls: Type[AsyncStreamT],
+    ) -> Union[ResponseT, AsyncStreamT]:
+        request = await self._build_request(options)
+        try:
+            res = await self._client.send(request, auth=self.custom_auth, stream=stream)
+            res.raise_for_status()
+        except httpx.HTTPStatusError as err:  # 4xx and 5xx errors
+            # If the response is streamed then we need to explicitly read the response
+            # to completion before attempting to access the response text.
+            err.response.read()
```

Comment:
This might have to be changed to its async version
`await err.response.aread()` other wise it raises an exception when the API responds with an error code